language/vm/src/file_format.rs

```diff
@@ -412,29 +417,25 @@ impl LocalsSignature {
 /// type parameter in the `FunctionSignature/Handle` and `StructHandle`.
 pub type TypeParameterIndex = u16;
 
-/// A `Kind` is the type of a type. It classifies types into categories with rules each category
-/// must follow.
+/// A `Kind` classifies types into sets with rules each set must follow.
 ///
-/// Currently there are two kinds in Move: `resource` and `copyable`.
+/// Currently there are three kinds in Move: `All`, `Resource` and `Unrestricted`.
 #[derive(Debug, Clone, Eq, Copy, Hash, Ord, PartialEq, PartialOrd)]
 #[cfg_attr(any(test, feature = "testing"), derive(Arbitrary))]
 pub enum Kind {
-    /// `resource` types must follow move semantics and various resource safety rules.
+    /// Represents the super set of all types. The type might actually be a `Resource` or
+    /// `Unrestricted` A type might be in this set if it is not known to be a `Resource` or
+    /// `Unrestricted`
+    ///   - This occurs when there is a type parameter with this kind as a constraint
+    All,
+    /// `Resource` types must follow move semantics and various resource safety rules, namely:
+    /// - `Resource` values cannot be copied
+    /// - `Resource` values cannot be popped, i.e. they must be used
     Resource,
-    /// `copyable` types do not need to follow the said rules. Most notably they can be freely
-    /// copied & destroyed. A `copyable` can still be used as a `resource` and therefore it is
-    /// considered a sub-kind of the latter.
-    Copyable,
-}
-
-impl Kind {
-    /// Checks if the kind is resource.
-    pub fn is_resource(self) -> bool {
-        match self {
-            Kind::Resource => true,
-            Kind::Copyable => false,
-        }
-    }
+    /// `Unrestricted` types do not need to follow the `Resource` rules.
+    /// - `Unrestricted` values can be copied
+    /// - `Unrestricted` values can be popped
+    Unrestricted,
```

Comment:
I like the intuition around `Copyable` vs `Unrestricted`, and `Resource` and `Unrestricted` sound good to me. Would `Generic` be a good name for the other type? I like it more than `All` or `Unknown` though I understand the debate it may lead to.
How is the syntax for generic type going to be in the IR? I ask as I am sure you have thought about it and it seems neither `struct` or `resource` are exact

Reply:
I think `Generic` would be a bit confusing with Generics, since the two concepts aren't really semantically related. 

The syntax of the IR will not change, since `is_nominal_resource` is just a renaming of the old `kind` field

Reply:
I was actually asking how is a generic declaration made in the IR, or will be made. I assume I can say
`resource R<T> {}`
`struct S<T> {}`
and for a T that is `All` or `Unknown` the syntax is `struct`?

Reply:
Correct
`struct S<T1, T2: resource, T3: unstricted>`
The struct handle for `S` would have `type_parameters`: `vec![All, Resource, Unrestricted]`